### PR TITLE
Replay route with history player

### PR DESCRIPTION
**Pull request by kmadsen**
## Description

Create parity with a new interface. The new route player should behave the same as the old replay route location engine.

- [x] I have added any issue links
- [ ] I have added all related labels (`bug`, `feature`, `new API(s)`, `SEMVER`, etc.)
- [ ] I have added the appropriate milestone and project boards

## Screenshots or Gifs

| old player | new player |
| -- | -- |
| output | output

## Testing

Please describe the manual tests that you ran to verify your changes

- [ ] I have tested locally (including `SNAPSHOT` upstream dependencies if needed) through testapp/demo app and run all activities to avoid regressions
- [x] I have tested via a test drive, or a simulation/mock location app
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

## Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have updated the `CHANGELOG` including this PR

### CODE REVIEW

**Comment by kmadsen** [comment]
> One thing that I noticed when testing is that it seems we're wrongly calculating the last locations in the route 👀

Yeah I made this work the same way as the existing replay engine. So the behavior is unchanged. Can get started on improving it

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/history/ReplayRouteMapper.kt [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+import android.location.Location
+import com.mapbox.api.directions.v5.models.DirectionsRoute
+import com.mapbox.geojson.LineString
+import com.mapbox.geojson.Point
+import com.mapbox.turf.TurfConstants
+import com.mapbox.turf.TurfMeasurement
+import java.util.Collections
+
+class ReplayRouteMapper(
```
Comment (RingerJK) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/history/ReplayRouteMapper.kt:11]: please add docs for public classes and methods

**Comment by kmadsen** [comment]
I am finding that simulated locations are not working well with enhanced locations. Building GPS simulated locations from a route will have a lot of effort.

Here is the video after updating the LocationComponent with RawLocations from the history player @Guardiola31337 

output

**Comment by Guardiola31337** [comment]
@kmadsen 
> I am finding that simulated locations are not working well with enhanced locations.

Not sure about this as a workaround 👀 

replay_route_history_player_cutting_corners

- Clients will need to add logic on their side depending on the `LocationEngine` used `onRawLocationChanged` vs `onEnhancedLocationChanged`
- Cutting corners when in simulation would regress 👀 https://github.com/mapbox/mapbox-navigation-android/pull/2452
- We're still seeing unexpected off-routes https://github.com/mapbox/mapbox-navigation-android/issues/2717

**Comment by kmadsen** [comment]
> Not sure about this as a workaround 👀

If we're generating locations, I think we should be controlling the view directly. We can rely on the history player to emulate GPS and real data.

Otherwise, will need some more time to see if building a location generator that is able to slow down at corners.

**Comment by kmadsen** [comment]
Closing this as it's going to go back in progress

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/history/ReplayRouteMapper.kt (hunk 1 above)
Comment (Guardiola31337) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/history/ReplayRouteMapper.kt:11]: We're missing multi-leg support to have parity with current `ReplayRouteLocationEngine` / `ReplayRouteLocationConverter` 👀 

https://github.com/mapbox/mapbox-navigation-android/blob/3044db38b6d26cd42fe69d7fc3544e1f81add612/libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/route/ReplayRouteLocationEngine.kt#L116 https://github.com/mapbox/mapbox-navigation-android/blob/3044db38b6d26cd42fe69d7fc3544e1f81add612/libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/route/ReplayRouteLocationConverter.kt#L21-L24

Also was wondering if we could re-use some code across `replay.history` and `replay.route` until `route` is replaced completely by `history`. What do you think?

Comment (kmadsen) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/history/ReplayRouteMapper.kt:11]: good catch, where are some examples of multi-leg? 

I'll see what it does when adding some tests

Comment (kmadsen) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/history/ReplayRouteMapper.kt:11]: Wondering if we should do this now or after we complete this
https://github.com/mapbox/navigation-sdks/issues/315

Comment (Guardiola31337) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/history/ReplayRouteMapper.kt:11]: Probably after :sweat_smile:

Comment (kmadsen) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/history/ReplayRouteMapper.kt:15]: Multiple initiatives to move forward waypoint support
https://github.com/mapbox/mapbox-navigation-android/pull/2786
https://github.com/mapbox/mapbox-navigation-android/pull/2787

making it so we can iterate on these things separately

Comment (kmadsen) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/history/ReplayRouteMapper.kt:15]: also docs added 🎈

File: examples/src/main/java/com/mapbox/navigation/examples/core/ReplayActivity.kt [hunk 2]
```diff
@@ -39,17 +39,14 @@ import timber.log.Timber
...
 class ReplayActivity : AppCompatActivity(), OnMapReadyCallback {
 
-    companion object {
-        const val DEFAULT_INTERVAL_IN_MILLISECONDS = 1000L
-        const val DEFAULT_MAX_WAIT_TIME = DEFAULT_INTERVAL_IN_MILLISECONDS * 5
-    }
-
-    private var mapboxMap: MapboxMap? = null
-    private var locationEngine: LocationEngine? = null
+    internal var mapboxMap: MapboxMap? = null
```
Comment (Guardiola31337) [inline examples/src/main/java/com/mapbox/navigation/examples/core/ReplayActivity.kt:42]:
```suggestion
    private var mapboxMap: MapboxMap? = null
```

**Pull request merged**